library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,165 @@
+// Based on
+// https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs
+// by matthieu-m
+use crate::alloc::{self, Layout};
+use core::cmp;
+use core::fmt::{self, Debug, Formatter};
+use core::marker::{PhantomData, Unsize};
+use core::mem;
+use core::ops::{Deref, DerefMut};
+use core::ptr::Pointee;
+use core::ptr::{self, NonNull};
+
+/// ThinBox.
+///
+/// A thin pointer, regardless of T.
+#[unstable(feature = "thin_box", issue = "92791")]
+pub struct ThinBox<T: ?Sized + Pointee> {
+    ptr: WithHeader<T::Metadata>,
```

Comment:
`T` always implements `Pointee`, so it might be clearer to remove `Pointee` bounds (here and elsewhere) and just write uses of `T::Metadata` like `<T as Pointee>::Metadata`:

```suggestion
pub struct ThinBox<T: ?Sized> {
    ptr: WithHeader<<T as Pointee>::Metadata>,
```